lib/Agrammon/Web/Routes.pm6

```diff
@@ -305,36 +305,38 @@ sub api-routes (Str $schema, $ws) {
         }
         operation 'exportExcel', -> LoggedIn $user {
             request-body -> %params {
+                # prevent header injection
+                my $filename = "%params<datasetName>.subst(/<-[\w\s_.-]>/, '', :g).xlsx";
+                my $excel = $ws.get-excel-export($user, %params).to-blob;
                 response.append-header(
                     'Content-disposition',
-                    # prevent header injection
-                    "attachment; filename=%params<datasetName>.subst(/<-[\w_.-]>/, '', :g).xlsx"
+                    qq{attachment; filename="$filename"}
                 );
-                content 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
-                    $ws.get-excel-export($user, %params).to-blob;
-                 CATCH {
+                content 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', $excel;
+                CATCH {
                     default {
                         note "$_";
                         response.status = 500;
-                        content 'application/json', %( error => .message )
+                        content 'text/html', html-error( %( :error(.message), :$filename ) );
                     }
                 }
             }
         }
         operation 'exportPDF', -> LoggedIn $user {
             request-body -> %params {
+                my $pdf = $ws.get-pdf-export($user, %params);
+                # prevent header injection
+                my $filename = "%params<datasetName>.subst(/<-[\w\s_.-]>/, '', :g).pdf";
```

Comment:
Same comment, but now I noticed the duplication, maybe the filename calculation wants factoring out.

Reply:
Put it into a sub.